### ISSUE TITLE
whitespace control removes wrong whitespace

### ISSUE THREAD

**Issue by damianham** [TechMagister/liquid.cr#31]
It seems the whitespace control is removing whitespace from within the template statements (unexpected behaviour) rather than removing the whitespace created by the flow control (expected behaviour).

Given this data
``` 
fields = [ {type: "reference", name: "user"}, {type: "reference", name: "post"} ]
```
with this template
```
{%- for field in fields -%}
{%- if field.type == "reference" -%}
  belongs_to :{{field.name}}
{%- endif -%}
{%- endfor -%}
```
I would expect the output to be
```
belongs_to :user
belongs_to :post
```
but the actual output is
```
belongs_to :userbelongs_to :post
```

**Comment by docelic** [comment]
Hey, you may be correct, although I don't see an obvious reason why the final effect would be such.
I will check how this could be slightly modified to have the original/intended behavior.

For the short term, not to delay you on your project, you could remove some "-" and it would work.

(The current logic is: if "{-" is found, delete all whitespace from the end of the preceeding block. If "-}" is found, delete all whitespace from the start of the following block.)

**Comment by damianham** [comment]
What we want to get rid of is all the extra lines from the flow control.  If you can make it so that a newline that follows '-%}' is not output then that might fix it.  By deleting all whitespace from the following block we are getting output lines joined together.

**Comment by faustinoaq** [comment]
@damianham I already working on a fix for this, so we can have clean formatting on recipes before release amber v0.8.0 :wink:

**Comment by damianham** [comment]
@faustinoaq @docelic if the current whitespace control is for minification please rethink it.  HTML minification can be achieved with compression by the web server.  If you want to minify with the renderer I suggest an option to turn it on in the API.

**Comment by faustinoaq** [comment]
@damianham Yeah, no problem I'm just trying to achieve what we need. I testing it with recipes :wink:

**Comment by faustinoaq** [comment]
@docelic The original Liquid language in Ruby uses a different logic to remove whitespaces :sweat_smile: 

Ref: [liquid/lexer.rb](https://github.com/Shopify/liquid/blob/5980ddbfaec83d20335781eab03088143e353453/lib/liquid/lexer.rb#L22)

Ref: https://github.com/Shopify/liquid/issues/216

**Comment by faustinoaq** [comment]
I already trying to do something similar :sweat_smile:

**Comment by damianham** [comment]
We can get rid of the newlines with {% expression -%}  i.e. the opening brace does not have the '-' however the formatting is still off (indentation is not preserved) and code formatting checks will still fail.

**Comment by docelic** [comment]
Hey @damianham, @faustinoaq reminded me yesterday that we should handle this.

I've ran some tests and basically the issue here is coming from the fact that the blocks are inside a loop (if they would not be in a loop, the flow control would work fine). I did not test this behavior, and obviously the slight differences between Liquid.cr's and original Liquid's parser are visible here.

I am not sure whether we are allowed to simply not output \n after an expression because the parser should be as close as possible to Liquid's. (If that's how Liquid does it, please confirm that.)

I will go looking more in depth from here and will re-read @faustinoaq's references to original Liquid, and will see what solution could be applied.

**Comment by damianham** [comment]
@docelic I can appreciate the desire to stay as close to Liquid as possible but when we can show that doing so means that is causes problems for us then I think we should be prepared to diverge.  From my point of view the lines containing only Liquid.cr expressions e.g. control flow, whether loops or conditions, should not appear in the output and that means removing trailing '\n' as well as the expression.  If a Liquid.cr user expects '{%-' to remove all preceding whitespace and '-%}' to remove all trailing whitespace then you can retain that behaviour.  I don't think any user of Liquid.cr would expect empty lines in the output after removing the expression so for me;
```
# a line
{% an expression %}\n
# another line
```
should produce the output
```
# a line
# another line
```

**Comment by docelic** [comment]
@TechMagister do you have any input on the above suggestions?

**Comment by docelic** [comment]
@faustinoaq @damianham On the way of understanding full details of this issue, I have created a small test case in Ruby's Liquid.

In it, I have placed Damian's template text:

```rb
require "liquid"
ctx = Liquid::Context.new
ctx[ "fields"]= [ {type: "reference", name: "user"}, {type: "reference", name: "post"} ]

txt = %Q( 
{%- for field in fields -%}
{%- if field.type == :reference -%}
  belongs_to :x
{%- endif -%}
{%- endfor -%}
)

puts Liquid::Template.parse(txt).render! ctx
```

And the output of this template was:

"belongs_to :xbelongs_to :x"

This leads me to believe that Liquid and Liquid.cr behave in exactly the same way, and if a templating language as widely used as Liquid didn't think there was a bug in that behavior, we shouldn't either.

So, to get as close as possible to Damian's desired output, I've removed two of the "minus" chars (here's an example from Crystal directly):

```cr
ctx = Context.new
ctx.set "fields", [ {type: "reference", name: "user"}, {type: "reference", name:

txt = %Q( 
{%- for field in fields -%} 
{%- if field.type == "reference" %}
  belongs_to :{{field.name}}
{% endif -%} 
{%- endfor -%} 
)

template = Liquid::Parser.parse txt 
puts template.render(ctx)
```

This template now outputs:

```

  belongs_to :user

  belongs_to :post
```

Which is again identical to original Liquid behavior.

From here, there are two ways in which the output can look exactly like Damian wants it:

Option 1: Put the "{% endif -%}" line directly after belongs_to, without a newline:

```
  belongs_to :{{field.name}}{% endif -%} 
```

The output of this is exactly what is asked:

```
  belongs_to :user
  belongs_to :post
```

Or alternatively, option 2, if Damian does not want to do this for some reason, then simply the recipes code needs to do a gsub() on the rendered template to remove occurrences of multiple newlines in a row.

Please let me know if you do not agree with this assessment. Thanks!

**Comment by TechMagister** [comment]
I agree with @docelic , the behaviours of Liquid.cr and the original Liquid have to be identical.

**Comment by damianham** [comment]
I am sure that Liquid authors don't think there is a bug in this behavior because the intended output is HTML which ignores whitespace.  Since we are generating code it is a problem for us.
@doecelic I agree that the gsub is not a desirable option as that would cause other problems.  I don't like the idea that the endif has to be placed in a specific place and some minus chars have to exist and some removed - that will only be more confusing.  I think the best option is to add a new control sequence to Liquid.cr so that it remains 100% compatible with the original Liquid when rendering HTML.  Can we add something like '!%}' to Liquid.cr or some other sequence of chars to signal this special behaviour to remove the trailing newline ?

**Comment by docelic** [comment]
Hey @damianham I am trying to say that the current behavior is correct for both code and HTML, and it gives maximum flexibility and is fully correct. The behavior you propose would be a special case, and it would specifically prevent some combinations from working, when a person would like to control spacing exactly but couldn't due to automatic removal of some parts that would happen inside liquid.cr.

I can't comment on your preferences regarding placing the line on the same or on the next line. I can only assure you that placing it on the same line is a standard and expected way to do it.

**Comment by faustinoaq** [comment]
> Option 1: Put the "{% endif -%}" line directly after belongs_to, without a newline:
>   `belongs_to :{{field.name}}{% endif -%} `

Thanks @docelic ! @damianham This should be enough for us. I mean this's just an additional step to get right formatting, we're already doing some weird stuff on ECR to get right indented code in amber templates, so why not on amber recipes? :sweat_smile: 

> Can we add something like '!%}' to Liquid.cr or some other sequence of chars to signal this special behaviour to remove the trailing newline ?

No, I think this is not the right way, because this shard is a "liquid" template implementation, so this should preserve same behavior as original.

**Comment by faustinoaq** [comment]
@docelic I tried your option 1:

```cr
require "liquid"

ctx = Liquid::Context.new
ctx.set "fields", [{type: "reference", name: "user1"}, {type: "reference", name: "user2"}]

txt = <<-CODE
{%- for field in fields -%} 
{%- if field.type == "reference" %}
  belongs_to :{{field.name}}{%- endif -%} 
{%- endfor -%} 
CODE

template = Liquid::Parser.parse txt 
puts template.render(ctx)
```

and I got this:

```
➜ ~ crystal liquid_option_1.cr

  belongs_to :user1
  belongs_to :user2
➜ ~
```

There is still a new line at first line, so I tried this:

```cr
require "liquid"

ctx = Liquid::Context.new
ctx.set "fields", [{type: "reference", name: "user1"}, {type: "reference", name: "user2"}]

txt = <<-CODE
{% for field in fields -%} 
{% if field.type == "reference" %}belongs_to :{{field.name}}{% endif %} 
{% endfor -%} 
CODE

template = Liquid::Parser.parse txt 
puts template.render(ctx)
```

and this removed the new line :tada: 

```
➜ ~ crystal liquid_faustinos_option.cr
belongs_to :user1 
belongs_to :user2 
➜ ~ 
```

So, I think using `{% if field.type == "reference" %}belongs_to :{{field.name}}{% endif %} ` in one line is not that bad, @damianham WDYT?

**Comment by faustinoaq** [comment]
@docelic Thank you! I just updated all recipes to fix formatting! :tada: 

https://github.com/amberframework/recipes/pull/15 Travis is passing now :sparkles: :100: 

@damianham I think we can close this, WDYT?
